// @recallnet/chains/dist/index.js
// src/index.ts
import { defineChain } from "viem";
import { anvil, filecoinCalibration } from "viem/chains";
import {
  DEVNET_CHAIN_ID,
  DEVNET_EVM_RPC_URL,
  DEVNET_EXPLORER_URL,
  DEVNET_OBJECT_API_URL,
  LOCALNET_CHAIN_ID,
  LOCALNET_EVM_RPC_URL,
  LOCALNET_EVM_WS_URL,
  LOCALNET_EXPLORER_URL,
  LOCALNET_OBJECT_API_URL,
  TESTNET_CHAIN_ID,
  TESTNET_EVM_RPC_URL,
  TESTNET_EVM_WS_URL,
  TESTNET_EXPLORER_URL,
  TESTNET_OBJECT_API_URL,
  TESTNET_REGISTRAR_URL
} from "@recallnet/network-constants";
var testnet = defineChain({
  id: Number(TESTNET_CHAIN_ID),
  name: "Recall Testnet",
  fees: {
    baseFeeMultiplier: 120
  },
  nativeCurrency: {
    name: "Recall",
    symbol: "RECALL",
    decimals: 18
  },
  rpcUrls: {
    default: {
      http: [TESTNET_EVM_RPC_URL],
      webSocket: [TESTNET_EVM_WS_URL]
    }
  },
  blockExplorers: {
    default: {
      name: "Recall Testnet Explorer",
      url: TESTNET_EXPLORER_URL
    }
  }
});
var localnet = defineChain({
  id: Number(LOCALNET_CHAIN_ID),
  name: "Recall Localnet",
  fees: {
    baseFeeMultiplier: 120
  },
  nativeCurrency: {
    name: "Recall",
    symbol: "RECALL",
    decimals: 18
  },
  rpcUrls: {
    default: {
      http: [LOCALNET_EVM_RPC_URL],
      webSocket: [LOCALNET_EVM_WS_URL]
    }
  },
  blockExplorers: {
    default: {
      name: "Recall Localnet Explorer",
      url: LOCALNET_EXPLORER_URL
    }
  }
});
var devnet = defineChain({
  id: Number(DEVNET_CHAIN_ID),
  name: "Recall Devnet",
  fees: {
    baseFeeMultiplier: 120
  },
  nativeCurrency: {
    name: "Recall",
    symbol: "RECALL",
    decimals: 18
  },
  rpcUrls: {
    default: {
      http: [DEVNET_EVM_RPC_URL]
    }
  },
  blockExplorers: {
    default: {
      name: "Recall Devnet Explorer",
      url: DEVNET_EXPLORER_URL
    }
  }
});
function supportedChains(isLocalDev = false) {
  const chains = [testnet, filecoinCalibration];
  if (isLocalDev) {
    chains.push(localnet, devnet, anvil);
  }
  chains.sort((a, b) => a.name > b.name ? 1 : -1);
  return chains;
}
function checkChainIsSupported(chain) {
  return supportedChains(true).some((c) => c.id === chain.id);
}
function getChain(chainIdOrName) {
  const chains = supportedChains(true);
  const chain = chains.find(
    (c) => c.id === chainIdOrName || c.name.toLowerCase().includes(chainIdOrName.toString().toLowerCase())
  );
  if (chain) {
    return chain;
  }
  throw new Error(`Chain ${chainIdOrName} not found`);
}
function checkHasParentChain(chain) {
  if (!checkChainIsSupported(chain))
    throw new Error(`Chain ${chain.name} not found`);
  switch (chain.id) {
    case testnet.id:
    case localnet.id:
      return true;
    default:
      return false;
  }
}
function getParentChain(chain) {
  switch (chain.id) {
    case testnet.id:
      return filecoinCalibration;
    case localnet.id:
      return anvil;
    default:
      return void 0;
  }
}
function checkChainName(chainName) {
  try {
    getChain(chainName);
    return true;
  } catch {
    return false;
  }
}
function getObjectApiUrl(chain) {
  switch (chain.id) {
    case testnet.id:
      return TESTNET_OBJECT_API_URL;
    case localnet.id:
      return LOCALNET_OBJECT_API_URL;
    case devnet.id:
      return DEVNET_OBJECT_API_URL;
    default:
      throw new Error(`Object API URL not found for chain ${chain.name}`);
  }
}
function getRegistrarUrl(chain) {
  switch (chain.id) {
    case testnet.id:
      return TESTNET_REGISTRAR_URL;
    default:
      throw new Error(`Registrar URL not found for chain ${chain.name}`);
  }
}
function getExplorerUrl(chain) {
  switch (chain.id) {
    case testnet.id:
      return TESTNET_EXPLORER_URL;
    case localnet.id:
      return LOCALNET_EXPLORER_URL;
    case devnet.id:
      return DEVNET_EXPLORER_URL;
    default:
      throw new Error(`Explorer URL not found for chain ${chain.name}`);
  }
}
export {
  checkChainIsSupported,
  checkChainName,
  checkHasParentChain,
  devnet,
  getChain,
  getExplorerUrl,
  getObjectApiUrl,
  getParentChain,
  getRegistrarUrl,
  localnet,
  supportedChains,
  testnet
};
