// @recallnet/network-constants/dist/index.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  DEVNET_CHAIN_ID: () => DEVNET_CHAIN_ID,
  DEVNET_EVM_RPC_URL: () => DEVNET_EVM_RPC_URL,
  DEVNET_EVM_WS_URL: () => DEVNET_EVM_WS_URL,
  DEVNET_EXPLORER_URL: () => DEVNET_EXPLORER_URL,
  DEVNET_GATEWAY_MANAGER_FACET_ADDRESS: () => DEVNET_GATEWAY_MANAGER_FACET_ADDRESS,
  DEVNET_OBJECT_API_URL: () => DEVNET_OBJECT_API_URL,
  DEVNET_RPC_URL: () => DEVNET_RPC_URL,
  DEVNET_SUBNET_GETTER_FACET_ADDRESS: () => DEVNET_SUBNET_GETTER_FACET_ADDRESS,
  DEVNET_SUBNET_ID: () => DEVNET_SUBNET_ID,
  LOCALNET_BLOB_MANAGER_ADDRESS: () => LOCALNET_BLOB_MANAGER_ADDRESS,
  LOCALNET_BUCKET_MANAGER_ADDRESS: () => LOCALNET_BUCKET_MANAGER_ADDRESS,
  LOCALNET_CHAIN_ID: () => LOCALNET_CHAIN_ID,
  LOCALNET_CREDIT_MANAGER_ADDRESS: () => LOCALNET_CREDIT_MANAGER_ADDRESS,
  LOCALNET_EVM_RPC_URL: () => LOCALNET_EVM_RPC_URL,
  LOCALNET_EVM_WS_URL: () => LOCALNET_EVM_WS_URL,
  LOCALNET_EXPLORER_URL: () => LOCALNET_EXPLORER_URL,
  LOCALNET_GATEWAY_MANAGER_FACET_ADDRESS: () => LOCALNET_GATEWAY_MANAGER_FACET_ADDRESS,
  LOCALNET_OBJECT_API_URL: () => LOCALNET_OBJECT_API_URL,
  LOCALNET_PARENT_CHAIN_ID: () => LOCALNET_PARENT_CHAIN_ID,
  LOCALNET_PARENT_ERC20_ADDRESS: () => LOCALNET_PARENT_ERC20_ADDRESS,
  LOCALNET_PARENT_EVM_RPC_URL: () => LOCALNET_PARENT_EVM_RPC_URL,
  LOCALNET_PARENT_GATEWAY_MANAGER_FACET_ADDRESS: () => LOCALNET_PARENT_GATEWAY_MANAGER_FACET_ADDRESS,
  LOCALNET_PARENT_SUBNET_GETTER_FACET_ADDRESS: () => LOCALNET_PARENT_SUBNET_GETTER_FACET_ADDRESS,
  LOCALNET_RPC_URL: () => LOCALNET_RPC_URL,
  LOCALNET_SUBNET_GETTER_FACET_ADDRESS: () => LOCALNET_SUBNET_GETTER_FACET_ADDRESS,
  LOCALNET_SUBNET_ID: () => LOCALNET_SUBNET_ID,
  MAX_OBJECT_SIZE: () => MAX_OBJECT_SIZE,
  MAX_QUERY_LIMIT: () => MAX_QUERY_LIMIT,
  MIN_TTL: () => MIN_TTL,
  RPC_TIMEOUT: () => RPC_TIMEOUT,
  TESTNET_BLOB_MANAGER_ADDRESS: () => TESTNET_BLOB_MANAGER_ADDRESS,
  TESTNET_BUCKET_MANAGER_ADDRESS: () => TESTNET_BUCKET_MANAGER_ADDRESS,
  TESTNET_CHAIN_ID: () => TESTNET_CHAIN_ID,
  TESTNET_CREDIT_MANAGER_ADDRESS: () => TESTNET_CREDIT_MANAGER_ADDRESS,
  TESTNET_EVM_RPC_URL: () => TESTNET_EVM_RPC_URL,
  TESTNET_EVM_WS_URL: () => TESTNET_EVM_WS_URL,
  TESTNET_EXPLORER_URL: () => TESTNET_EXPLORER_URL,
  TESTNET_FAUCET_URL: () => TESTNET_FAUCET_URL,
  TESTNET_GATEWAY_MANAGER_FACET_ADDRESS: () => TESTNET_GATEWAY_MANAGER_FACET_ADDRESS,
  TESTNET_OBJECT_API_URL: () => TESTNET_OBJECT_API_URL,
  TESTNET_PARENT_CHAIN_ID: () => TESTNET_PARENT_CHAIN_ID,
  TESTNET_PARENT_ERC20_ADDRESS: () => TESTNET_PARENT_ERC20_ADDRESS,
  TESTNET_PARENT_EVM_RPC_URL: () => TESTNET_PARENT_EVM_RPC_URL,
  TESTNET_PARENT_GATEWAY_MANAGER_FACET_ADDRESS: () => TESTNET_PARENT_GATEWAY_MANAGER_FACET_ADDRESS,
  TESTNET_PARENT_SUBNET_GETTER_FACET_ADDRESS: () => TESTNET_PARENT_SUBNET_GETTER_FACET_ADDRESS,
  TESTNET_REGISTRAR_URL: () => TESTNET_REGISTRAR_URL,
  TESTNET_RPC_URL: () => TESTNET_RPC_URL,
  TESTNET_SUBNET_GETTER_FACET_ADDRESS: () => TESTNET_SUBNET_GETTER_FACET_ADDRESS,
  TESTNET_SUBNET_ID: () => TESTNET_SUBNET_ID
});
module.exports = __toCommonJS(index_exports);
var TESTNET_SUBNET_ID = "/r314159/t410fntaew3dtef2zpfox2hlhy4cgrksdt5zjwg3hbfi";
var LOCALNET_SUBNET_ID = "/r31337/t410f6gbdxrbehnaeeo4mrq7wc5hgq6smnefys4qanwi";
var DEVNET_SUBNET_ID = "test";
var TESTNET_PARENT_CHAIN_ID = 314159;
var TESTNET_CHAIN_ID = 2481632;
var LOCALNET_PARENT_CHAIN_ID = 31337;
var LOCALNET_CHAIN_ID = 248163216;
var DEVNET_CHAIN_ID = 1942764459484029;
var TESTNET_RPC_URL = "https://api.testnet.recall.chain.love";
var LOCALNET_RPC_URL = "http://127.0.0.1:26657";
var DEVNET_RPC_URL = "http://127.0.0.1:26657";
var TESTNET_EVM_RPC_URL = "https://evm.testnet.recall.chain.love";
var TESTNET_EVM_WS_URL = "wss://evm.testnet.recall.chain.love";
var TESTNET_PARENT_EVM_RPC_URL = "https://api.calibration.node.glif.io/rpc/v1";
var LOCALNET_EVM_RPC_URL = "http://127.0.0.1:8645";
var LOCALNET_EVM_WS_URL = "ws://127.0.0.1:8645";
var LOCALNET_PARENT_EVM_RPC_URL = "http://127.0.0.1:8545";
var DEVNET_EVM_RPC_URL = "http://127.0.0.1:8545";
var DEVNET_EVM_WS_URL = "ws://127.0.0.1:8545";
var TESTNET_OBJECT_API_URL = "https://objects.testnet.recall.chain.love";
var LOCALNET_OBJECT_API_URL = "http://127.0.0.1:8001";
var DEVNET_OBJECT_API_URL = "http://127.0.0.1:8001";
var TESTNET_REGISTRAR_URL = "https://registrar.node-0.testnet.recall.network";
var TESTNET_EXPLORER_URL = "https://explorer.testnet.recall.network";
var LOCALNET_EXPLORER_URL = "http://127.0.0.1:8000";
var DEVNET_EXPLORER_URL = "http://127.0.0.1:8000";
var TESTNET_FAUCET_URL = "https://faucet.recall.network";
var RPC_TIMEOUT = 6e4;
var MIN_TTL = 3600n;
var MAX_OBJECT_SIZE = 5e9;
var MAX_QUERY_LIMIT = 50;
var TESTNET_PARENT_GATEWAY_MANAGER_FACET_ADDRESS = "0x2758f99EaB8ea9B8678B5d841851D62Ef18AAB26";
var TESTNET_GATEWAY_MANAGER_FACET_ADDRESS = "0x77aa40b105843728088c0132e43fc44348881da8";
var LOCALNET_PARENT_GATEWAY_MANAGER_FACET_ADDRESS = "0x9A676e781A523b5d0C0e43731313A708CB607508";
var LOCALNET_GATEWAY_MANAGER_FACET_ADDRESS = "0x77aa40b105843728088c0132e43fc44348881da8";
var DEVNET_GATEWAY_MANAGER_FACET_ADDRESS = "0x77aa40b105843728088c0132e43fc44348881da8";
var TESTNET_PARENT_SUBNET_GETTER_FACET_ADDRESS = "0xE8090d55E7ecc0565830845Df67bfD3fA81D4158";
var LOCALNET_PARENT_SUBNET_GETTER_FACET_ADDRESS = "0x322813Fd9A801c5507c9de605d63CEA4f2CE6c44";
var TESTNET_SUBNET_GETTER_FACET_ADDRESS = "0x74539671a1d2f1c8f200826baba665179f53a1b7";
var LOCALNET_SUBNET_GETTER_FACET_ADDRESS = "0x74539671a1d2f1c8f200826baba665179f53a1b7";
var DEVNET_SUBNET_GETTER_FACET_ADDRESS = "0x74539671a1d2f1c8f200826baba665179f53a1b7";
var TESTNET_PARENT_ERC20_ADDRESS = "0xd1239c6b6f806EC3752df12CEaddD88187BCf1E5";
var LOCALNET_PARENT_ERC20_ADDRESS = "0x4A679253410272dd5232B3Ff7cF5dbB88f295319";
var TESTNET_BLOB_MANAGER_ADDRESS = "0x27594F8412bA08278b15b55C6347605Ee3b2020C";
var LOCALNET_BLOB_MANAGER_ADDRESS = "0x8ce361602B935680E8DeC218b820ff5056BeB7af";
var TESTNET_BUCKET_MANAGER_ADDRESS = "0x170a22c76a2Bb7c5F1b0e7067a0ddb7C3452c8e0";
var LOCALNET_BUCKET_MANAGER_ADDRESS = "0xeD1DB453C3156Ff3155a97AD217b3087D5Dc5f6E";
var TESTNET_CREDIT_MANAGER_ADDRESS = "0x296042865b783D8E8Fb09F04FFAE2A66a16B85Bc";
var LOCALNET_CREDIT_MANAGER_ADDRESS = "0x196dBCBb54b8ec4958c959D8949EBFE87aC2Aaaf";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEVNET_CHAIN_ID,
  DEVNET_EVM_RPC_URL,
  DEVNET_EVM_WS_URL,
  DEVNET_EXPLORER_URL,
  DEVNET_GATEWAY_MANAGER_FACET_ADDRESS,
  DEVNET_OBJECT_API_URL,
  DEVNET_RPC_URL,
  DEVNET_SUBNET_GETTER_FACET_ADDRESS,
  DEVNET_SUBNET_ID,
  LOCALNET_BLOB_MANAGER_ADDRESS,
  LOCALNET_BUCKET_MANAGER_ADDRESS,
  LOCALNET_CHAIN_ID,
  LOCALNET_CREDIT_MANAGER_ADDRESS,
  LOCALNET_EVM_RPC_URL,
  LOCALNET_EVM_WS_URL,
  LOCALNET_EXPLORER_URL,
  LOCALNET_GATEWAY_MANAGER_FACET_ADDRESS,
  LOCALNET_OBJECT_API_URL,
  LOCALNET_PARENT_CHAIN_ID,
  LOCALNET_PARENT_ERC20_ADDRESS,
  LOCALNET_PARENT_EVM_RPC_URL,
  LOCALNET_PARENT_GATEWAY_MANAGER_FACET_ADDRESS,
  LOCALNET_PARENT_SUBNET_GETTER_FACET_ADDRESS,
  LOCALNET_RPC_URL,
  LOCALNET_SUBNET_GETTER_FACET_ADDRESS,
  LOCALNET_SUBNET_ID,
  MAX_OBJECT_SIZE,
  MAX_QUERY_LIMIT,
  MIN_TTL,
  RPC_TIMEOUT,
  TESTNET_BLOB_MANAGER_ADDRESS,
  TESTNET_BUCKET_MANAGER_ADDRESS,
  TESTNET_CHAIN_ID,
  TESTNET_CREDIT_MANAGER_ADDRESS,
  TESTNET_EVM_RPC_URL,
  TESTNET_EVM_WS_URL,
  TESTNET_EXPLORER_URL,
  TESTNET_FAUCET_URL,
  TESTNET_GATEWAY_MANAGER_FACET_ADDRESS,
  TESTNET_OBJECT_API_URL,
  TESTNET_PARENT_CHAIN_ID,
  TESTNET_PARENT_ERC20_ADDRESS,
  TESTNET_PARENT_EVM_RPC_URL,
  TESTNET_PARENT_GATEWAY_MANAGER_FACET_ADDRESS,
  TESTNET_PARENT_SUBNET_GETTER_FACET_ADDRESS,
  TESTNET_REGISTRAR_URL,
  TESTNET_RPC_URL,
  TESTNET_SUBNET_GETTER_FACET_ADDRESS,
  TESTNET_SUBNET_ID
});
