// @recallnet/fvm/dist/artifacts.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/artifacts/index.ts
var artifacts_exports = {};
__export(artifacts_exports, {
  ACTOR_ID_ETHEREUM_MASK: () => ACTOR_ID_ETHEREUM_MASK,
  ACTOR_ID_ETHEREUM_MASK_LEN: () => ACTOR_ID_ETHEREUM_MASK_LEN,
  ACTOR_PAYLOAD_LEN: () => ACTOR_PAYLOAD_LEN,
  ActorsCalibrationV11: () => ActorsCalibrationV11,
  ActorsCalibrationV12: () => ActorsCalibrationV12,
  ActorsMainnetV10: () => ActorsMainnetV10,
  ActorsSubnet: () => ActorsSubnet,
  BLS_PAYLOAD_LEN: () => BLS_PAYLOAD_LEN,
  DelegatedNamespace: () => DelegatedNamespace,
  ETH_ADDRESS_LEN: () => ETH_ADDRESS_LEN,
  ID_PAYLOAD_MAX_NUM: () => ID_PAYLOAD_MAX_NUM,
  InitActorMethods: () => InitActorMethods,
  Network: () => Network,
  NetworkPrefix: () => NetworkPrefix,
  PayChActorMethods: () => PayChActorMethods,
  ProtocolIndicator: () => ProtocolIndicator,
  SECP256K1_PAYLOAD_LEN: () => SECP256K1_PAYLOAD_LEN,
  SUB_ADDRESS_MAX_LEN: () => SUB_ADDRESS_MAX_LEN,
  SystemActorIDs: () => SystemActorIDs
});
module.exports = __toCommonJS(artifacts_exports);

// src/artifacts/actors.ts
var SystemActorIDs = /* @__PURE__ */ ((SystemActorIDs2) => {
  SystemActorIDs2[SystemActorIDs2["System"] = 0] = "System";
  SystemActorIDs2[SystemActorIDs2["Init"] = 1] = "Init";
  SystemActorIDs2[SystemActorIDs2["Reward"] = 2] = "Reward";
  SystemActorIDs2[SystemActorIDs2["Cron"] = 3] = "Cron";
  SystemActorIDs2[SystemActorIDs2["StoragePower"] = 4] = "StoragePower";
  SystemActorIDs2[SystemActorIDs2["StorageMarket"] = 5] = "StorageMarket";
  SystemActorIDs2[SystemActorIDs2["VerifiedRegistry"] = 6] = "VerifiedRegistry";
  SystemActorIDs2[SystemActorIDs2["DataCap"] = 7] = "DataCap";
  SystemActorIDs2[SystemActorIDs2["EAM"] = 10] = "EAM";
  SystemActorIDs2[SystemActorIDs2["ADM"] = 17] = "ADM";
  return SystemActorIDs2;
})(SystemActorIDs || {});
var InitActorMethods = /* @__PURE__ */ ((InitActorMethods2) => {
  InitActorMethods2[InitActorMethods2["Exec"] = 2] = "Exec";
  InitActorMethods2[InitActorMethods2["Exec4"] = 3] = "Exec4";
  return InitActorMethods2;
})(InitActorMethods || {});
var PayChActorMethods = /* @__PURE__ */ ((PayChActorMethods2) => {
  PayChActorMethods2[PayChActorMethods2["UpdateChannelState"] = 2] = "UpdateChannelState";
  PayChActorMethods2[PayChActorMethods2["Settle"] = 3] = "Settle";
  PayChActorMethods2[PayChActorMethods2["Collect"] = 4] = "Collect";
  return PayChActorMethods2;
})(PayChActorMethods || {});
var ActorsCalibrationV12 = /* @__PURE__ */ ((ActorsCalibrationV122) => {
  ActorsCalibrationV122["Account"] = "bafk2bzacechwwxdqvggkdylm37zldjsra2ivkdzwp7fee56bzxbzs544wv6u6";
  ActorsCalibrationV122["Cron"] = "bafk2bzacec4gdxxkqwxqqodsv6ug5dmdbqdfqwyqfek3yhxc2wweh5psxaeq6";
  ActorsCalibrationV122["DataCap"] = "bafk2bzacecq5ppfskxgv3iea3jarsix6jdduuhwsn4fbvngtbmzelzmlygorm";
  ActorsCalibrationV122["EAM"] = "bafk2bzacecb6cnwftvavpph4p34zs4psuy5xvbrhf7vszkva4npw6mw3c42xe";
  ActorsCalibrationV122["EthAccount"] = "bafk2bzaceajmc3y3sedsqymfla3dzzqzmbu5kmr2iskm26ga2u34ll5fpztfw";
  ActorsCalibrationV122["EVM"] = "bafk2bzaced4sozr7m6rzcgpobzeiupghthfw6afumysu3oz6bxxirv74uo3vw";
  ActorsCalibrationV122["Init"] = "bafk2bzaceaewh7b6zl2egclm7fqzx2lsqr57i75lb6cj43ndoa4mal3k5ld3m";
  ActorsCalibrationV122["Multisig"] = "bafk2bzacednkwcpw5yzxjceoaliajgupzj6iqxe7ks2ll3unspbprbo5f2now";
  ActorsCalibrationV122["PaymentChannel"] = "bafk2bzacebaxhk4itfiuvbftg7kz5zxugqnvdgerobitjq4vl6q4orcwk6wqg";
  ActorsCalibrationV122["Placeholder"] = "bafk2bzacedfvut2myeleyq67fljcrw4kkmn5pb5dpyozovj7jpoez5irnc3ro";
  ActorsCalibrationV122["Reward"] = "bafk2bzacedra77pcglf7vdca2itcaa4vd6xrxynxmgfgdjdxqxfwqyhtoxehy";
  ActorsCalibrationV122["StorageMarket"] = "bafk2bzacea7g46y7xxu2zjq2h75x6mmx3utz2uxnlvnwi6tzpsvulna3bmiva";
  ActorsCalibrationV122["StorageMiner"] = "bafk2bzaced7emkbbnrewv5uvrokxpf5tlm4jslu2jsv77ofw2yqdglg657uie";
  ActorsCalibrationV122["StoragePower"] = "bafk2bzacedd3ka44k7d46ckbinjhv3diyuu2epgbyvhqqyjkc64qlrg3wlgzi";
  ActorsCalibrationV122["System"] = "bafk2bzacecioupndtcnyw6iq2hbrxag3aufvczlv5nobnfbkbywqzcyfaa376";
  ActorsCalibrationV122["VerifiedRegistry"] = "bafk2bzaceavldupmf7bimeeacs67z5xdfdlfca6p7sn6bev3mt5ggepfqvhqo";
  return ActorsCalibrationV122;
})(ActorsCalibrationV12 || {});
var ActorsCalibrationV11 = /* @__PURE__ */ ((ActorsCalibrationV112) => {
  ActorsCalibrationV112["Account"] = "bafk2bzacebor5mnjnsav34cmm5pcd3dy4wubbv4wtcrvba7depy3sct7ie4sy";
  ActorsCalibrationV112["Cron"] = "bafk2bzacebetehhedh55alfn4rcx2mhjhvuiustxlhtxc3drkemnpttws5eqw";
  ActorsCalibrationV112["DataCap"] = "bafk2bzaced6uhmrh5jjexhw4lco4ipesi2iutl7uupnyspgmnbydyo3amtu4i";
  ActorsCalibrationV112["EAM"] = "bafk2bzacea6wzcnflfnaxqnwydoghh7ezg5au32ew3bnzljzpiw6fimhlpoiu";
  ActorsCalibrationV112["EthAccount"] = "bafk2bzacedrbpvjvyzif2cjxosm4pliyq2m6wzndvrg7r6hzdhixplzvgubbw";
  ActorsCalibrationV112["EVM"] = "bafk2bzaceabftmhejmvjvpzmbsv4cvaew6v5juj5sqtq7cfijugwsnahnsy5w";
  ActorsCalibrationV112["Init"] = "bafk2bzaceduyjd35y7o2lhvevtysqf45rp5ot7x5f36q6iond6dyiz6773g5q";
  ActorsCalibrationV112["Multisig"] = "bafk2bzacebcb72fmbpocetnzgni2wnbrduamlqx6fl3yelrlzu7id6bu5ib5g";
  ActorsCalibrationV112["PaymentChannel"] = "bafk2bzaceazwhm63kyp47pste5i5acnuhosrgythyagf3kc5clogiqqx6vkzk";
  ActorsCalibrationV112["Placeholder"] = "bafk2bzacedfvut2myeleyq67fljcrw4kkmn5pb5dpyozovj7jpoez5irnc3ro";
  ActorsCalibrationV112["Reward"] = "bafk2bzacecp7xo5ev46y64zr5osnn5fxo7itpoqw235tcfv6eo4rymzdemet2";
  ActorsCalibrationV112["StorageMarket"] = "bafk2bzacedjt5mueomasx7dijooxnwxsbtzu2dj2ppp45rtle4kiinkmgzeei";
  ActorsCalibrationV112["StorageMiner"] = "bafk2bzacebkjnjp5okqjhjxzft5qkuv36u4tz7inawseiwi2kw4j43xpxvhpm";
  ActorsCalibrationV112["StoragePower"] = "bafk2bzaced2qsypqwore3jrdtaesh4itst2fyeepdsozvtffc2pianzmphdum";
  ActorsCalibrationV112["System"] = "bafk2bzacedqvik2n3phnj3cni3h2k5mtvz43nyq7mdmv7k7euejysvajywdug";
  ActorsCalibrationV112["VerifiedRegistry"] = "bafk2bzaceceoo5jlom2zweh7kpye2vkj33wgqnkjshlsw2neemqkfg5g2rmvg";
  return ActorsCalibrationV112;
})(ActorsCalibrationV11 || {});
var ActorsMainnetV10 = /* @__PURE__ */ ((ActorsMainnetV102) => {
  ActorsMainnetV102["Account"] = "bafk2bzaceampw4romta75hyz5p4cqriypmpbgnkxncgxgqn6zptv5lsp2w2bo";
  ActorsMainnetV102["Cron"] = "bafk2bzacedcbtsifegiu432m5tysjzkxkmoczxscb6hqpmrr6img7xzdbbs2g";
  ActorsMainnetV102["DataCap"] = "bafk2bzacealj5uk7wixhvk7l5tnredtelralwnceafqq34nb2lbylhtuyo64u";
  ActorsMainnetV102["EAM"] = "bafk2bzacedrpm5gbleh4xkyo2jvs7p5g6f34soa6dpv7ashcdgy676snsum6g";
  ActorsMainnetV102["EthAccount"] = "bafk2bzaceaqoc5zakbhjxn3jljc4lxnthllzunhdor7sxhwgmskvc6drqc3fa";
  ActorsMainnetV102["EVM"] = "bafk2bzaceahmzdxhqsm7cu2mexusjp6frm7r4kdesvti3etv5evfqboos2j4g";
  ActorsMainnetV102["Init"] = "bafk2bzaced2f5rhir3hbpqbz5ght7ohv2kgj42g5ykxrypuo2opxsup3ykwl6";
  ActorsMainnetV102["Multisig"] = "bafk2bzaceduf3hayh63jnl4z2knxv7cnrdenoubni22fxersc4octlwpxpmy4";
  ActorsMainnetV102["PaymentChannel"] = "bafk2bzaceartlg4mrbwgzcwric6mtvyawpbgx2xclo2vj27nna57nxynf3pgc";
  ActorsMainnetV102["Placeholder"] = "bafk2bzacedfvut2myeleyq67fljcrw4kkmn5pb5dpyozovj7jpoez5irnc3ro";
  ActorsMainnetV102["Reward"] = "bafk2bzacebnhtaejfjtzymyfmbdrfmo7vgj3zsof6zlucbmkhrvcuotw5dxpq";
  ActorsMainnetV102["StorageMarket"] = "bafk2bzaceclejwjtpu2dhw3qbx6ow7b4pmhwa7ocrbbiqwp36sq5yeg6jz2bc";
  ActorsMainnetV102["StorageMiner"] = "bafk2bzaced4h7noksockro7glnssz2jnmo2rpzd7dvnmfs4p24zx3h6gtx47s";
  ActorsMainnetV102["StoragePower"] = "bafk2bzacec4ay4crzo73ypmh7o3fjendhbqrxake46bprabw67fvwjz5q6ixq";
  ActorsMainnetV102["System"] = "bafk2bzacedakk5nofebyup4m7nvx6djksfwhnxzrfuq4oyemhpl4lllaikr64";
  ActorsMainnetV102["VerifiedRegistry"] = "bafk2bzacedfel6edzqpe5oujno7fog4i526go4dtcs6vwrdtbpy2xq6htvcg6";
  return ActorsMainnetV102;
})(ActorsMainnetV10 || {});
var ActorsSubnet = /* @__PURE__ */ ((ActorsSubnet2) => {
  ActorsSubnet2["Account"] = "bafk2bzacecomydaym5xi2burqfxv3kz2xjr3vgpig67fm6gjvqzfxre3z4wku";
  ActorsSubnet2["Cron"] = "bafk2bzaceagm5m56lzzue5bpjtxvotsnikjigsjyk2mmrhjna35vkbliaellg";
  ActorsSubnet2["DataCap"] = "bafk2bzacebdwzssyouoz6mfydy72c26euw3upec6nxtwme23bjcbsms6fppei";
  ActorsSubnet2["EAM"] = "bafk2bzaceadwtvjqlvrgz3pruas47offhoqtxk5oo3ptigp5ujw4szhepletq";
  ActorsSubnet2["EthAccount"] = "bafk2bzacebrf6ijl3tdtqgrnv5zwmcnvm5lfwjgpr5lnt42lolt2gmvpuluas";
  ActorsSubnet2["EVM"] = "bafk2bzaceckc2hwgiarq7rjmyefdsm7qbv4ybr2wmt52amihgbomzjg23qta6";
  ActorsSubnet2["Init"] = "bafk2bzacebgrvihpo23xutjv3hgoxlq55z2gx5qjwqwmu2xytp7tun2wgvgsw";
  ActorsSubnet2["Multisig"] = "bafk2bzacea7c2c2nf3a6ksfs7buxynugwjcc7r3fuirjf4wwsy5baqyfacqny";
  ActorsSubnet2["PaymentChannel"] = "bafk2bzaceblv6iudi6dr42lihhf46gxedzfinrc5udktxxxjiwbdscazxkieg";
  ActorsSubnet2["Placeholder"] = "bafk2bzacedfvut2myeleyq67fljcrw4kkmn5pb5dpyozovj7jpoez5irnc3ro";
  ActorsSubnet2["Reward"] = "bafk2bzacechfthpzob7b7u27zhzpv3q6axtv27qkgtzgcokukl3hhmqpbqv56";
  ActorsSubnet2["StorageMarket"] = "bafk2bzacebrc3arb6y2udwbxtambzbewl2fq3l5zguxeihisfkc4fyoghj44q";
  ActorsSubnet2["StorageMiner"] = "bafk2bzacecdlrg32cpbbnnoselcx4y4ottxwdcm2vhs4eropvaxrulmvcrijw";
  ActorsSubnet2["StoragePower"] = "bafk2bzaced3m5k5wnvnkq4wjhaalwspqa3jpn7ok4qv4jwxlfp7jjovvjk3he";
  ActorsSubnet2["System"] = "bafk2bzaceb7an3qpibd2x3a4k2ng4xkgqcbsweaefjpbhxznokxmqij3kyrmy";
  ActorsSubnet2["VerifiedRegistry"] = "bafk2bzacedvd4gp2v2qxcs6wohlur4hio2mxs2kvcyzysg66xmrddlmcnkavm";
  ActorsSubnet2["ADM"] = "bafk2bzaceb3un4rdw24kbhokgs7vqulh4ptoaua6ynkpgjsbffy7qb5lkj65w";
  return ActorsSubnet2;
})(ActorsSubnet || {});

// src/artifacts/address.ts
var DelegatedNamespace = /* @__PURE__ */ ((DelegatedNamespace2) => {
  DelegatedNamespace2["ETH"] = "10";
  return DelegatedNamespace2;
})(DelegatedNamespace || {});
var ACTOR_ID_ETHEREUM_MASK = 255;
var ACTOR_ID_ETHEREUM_MASK_LEN = 12;
var ProtocolIndicator = /* @__PURE__ */ ((ProtocolIndicator2) => {
  ProtocolIndicator2[ProtocolIndicator2["ID"] = 0] = "ID";
  ProtocolIndicator2[ProtocolIndicator2["SECP256K1"] = 1] = "SECP256K1";
  ProtocolIndicator2[ProtocolIndicator2["ACTOR"] = 2] = "ACTOR";
  ProtocolIndicator2[ProtocolIndicator2["BLS"] = 3] = "BLS";
  ProtocolIndicator2[ProtocolIndicator2["DELEGATED"] = 4] = "DELEGATED";
  return ProtocolIndicator2;
})(ProtocolIndicator || {});
var Network = /* @__PURE__ */ ((Network2) => {
  Network2["Mainnet"] = "mainnet";
  Network2["Testnet"] = "testnet";
  return Network2;
})(Network || {});
var NetworkPrefix = /* @__PURE__ */ ((NetworkPrefix2) => {
  NetworkPrefix2["Mainnet"] = "f";
  NetworkPrefix2["Testnet"] = "t";
  return NetworkPrefix2;
})(NetworkPrefix || {});
var ETH_ADDRESS_LEN = 20;
var SUB_ADDRESS_MAX_LEN = 54;
var BLS_PAYLOAD_LEN = 48;
var ACTOR_PAYLOAD_LEN = 20;
var SECP256K1_PAYLOAD_LEN = 20;
var ID_PAYLOAD_MAX_NUM = 2n ** 63n - 1n;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ACTOR_ID_ETHEREUM_MASK,
  ACTOR_ID_ETHEREUM_MASK_LEN,
  ACTOR_PAYLOAD_LEN,
  ActorsCalibrationV11,
  ActorsCalibrationV12,
  ActorsMainnetV10,
  ActorsSubnet,
  BLS_PAYLOAD_LEN,
  DelegatedNamespace,
  ETH_ADDRESS_LEN,
  ID_PAYLOAD_MAX_NUM,
  InitActorMethods,
  Network,
  NetworkPrefix,
  PayChActorMethods,
  ProtocolIndicator,
  SECP256K1_PAYLOAD_LEN,
  SUB_ADDRESS_MAX_LEN,
  SystemActorIDs
});
