// @recallnet/fvm/dist/chunk-P4BZGKN5.mjs
// src/artifacts/address.ts
var DelegatedNamespace = /* @__PURE__ */ ((DelegatedNamespace2) => {
  DelegatedNamespace2["ETH"] = "10";
  return DelegatedNamespace2;
})(DelegatedNamespace || {});
var ACTOR_ID_ETHEREUM_MASK = 255;
var ACTOR_ID_ETHEREUM_MASK_LEN = 12;
var ProtocolIndicator = /* @__PURE__ */ ((ProtocolIndicator2) => {
  ProtocolIndicator2[ProtocolIndicator2["ID"] = 0] = "ID";
  ProtocolIndicator2[ProtocolIndicator2["SECP256K1"] = 1] = "SECP256K1";
  ProtocolIndicator2[ProtocolIndicator2["ACTOR"] = 2] = "ACTOR";
  ProtocolIndicator2[ProtocolIndicator2["BLS"] = 3] = "BLS";
  ProtocolIndicator2[ProtocolIndicator2["DELEGATED"] = 4] = "DELEGATED";
  return ProtocolIndicator2;
})(ProtocolIndicator || {});
var Network = /* @__PURE__ */ ((Network2) => {
  Network2["Mainnet"] = "mainnet";
  Network2["Testnet"] = "testnet";
  return Network2;
})(Network || {});
var NetworkPrefix = /* @__PURE__ */ ((NetworkPrefix2) => {
  NetworkPrefix2["Mainnet"] = "f";
  NetworkPrefix2["Testnet"] = "t";
  return NetworkPrefix2;
})(NetworkPrefix || {});
var ETH_ADDRESS_LEN = 20;
var SUB_ADDRESS_MAX_LEN = 54;
var BLS_PAYLOAD_LEN = 48;
var ACTOR_PAYLOAD_LEN = 20;
var SECP256K1_PAYLOAD_LEN = 20;
var ID_PAYLOAD_MAX_NUM = 2n ** 63n - 1n;

export {
  DelegatedNamespace,
  ACTOR_ID_ETHEREUM_MASK,
  ACTOR_ID_ETHEREUM_MASK_LEN,
  ProtocolIndicator,
  Network,
  NetworkPrefix,
  ETH_ADDRESS_LEN,
  SUB_ADDRESS_MAX_LEN,
  BLS_PAYLOAD_LEN,
  ACTOR_PAYLOAD_LEN,
  SECP256K1_PAYLOAD_LEN,
  ID_PAYLOAD_MAX_NUM
};
